modules/internal/DefaultAssociative.chpl

```diff
@@ -106,16 +106,81 @@ module DefaultAssociative {
     }
   
     proc dsiSerialReadWrite(f /*: Reader or Writer*/) {
-      var first = true;
-      f <~> new ioLiteral("{");
-      for idx in this {
-        if first then 
-          first = false; 
-        else 
-          f <~> new ioLiteral(", ");
-        f <~> idx;
-      }
-      f <~> new ioLiteral("}");
+
+      var binary = f.binary();
+
+      if f.writing {
+        if binary {
+          var numIndices: int = dsiNumIndices;
+          f <~> numIndices;
+          for idx in this {
+            f <~> idx;
+          }
+        } else {
+          var first = true;
+          f <~> new ioLiteral("{");
+          for idx in this {
+            if first then 
+              first = false; 
+            else 
+              f <~> new ioLiteral(", ");
+            f <~> idx;
+          }
+          f <~> new ioLiteral("}");
+        }
+      } else {
+        if binary {
+          var numIndices: int;
+          f <~> numIndices;
+          for i in 1..numIndices {
+            var idx: idxType;
+            f <~> idx;
+            dsiAdd(idx);
+          }
+        } else {
+          f <~> new ioLiteral("{");
+
+          var first = true;
+          var comma = new ioLiteral(",", true);
+          var end = new ioLiteral("}");
+
+          while true {
+            // Try reading an end curly
+            writeln("Reading end curly");
+            f.readwrite(end);
```

Comment:
Why use `f.readwrite()` instead of the `<~>` operator on this branch?  For consistency with `f.error()` and `f.clearError()` maybe?